### PR TITLE
rpm2img: remove core kit specifics, set kit vendor as Publisher

### PR DESCRIPTION
**Pull request by ginglis13**
**Issue number:**

Related: #397 

**Description of changes:**

Remove core kit specific logic from application inventory generation such that multiple source kits can be used.

Additionally, set Publisher to a kit's vendor for packages that come from kits.

**Testing done:**

Built a `k8s-1.24` variant:

<details open>
<summary>Application Inventory</summary>
<br>

https://gist.github.com/ginglis13/58828f4c62242a53cbf139b1f9a5a307

```
...
 {                                                                                                                                                                                              18:30:25 [207/7752]
      "Name": "bottlerocket-settings-plugins-aws-k8s",
      "Publisher": "Bottlerocket",
      "Version": "0.0",
      "Release": "1.1731974957.b2bc1ccf.br1",
      "Epoch": "0",
      "InstalledTime": "2024-11-22T18:29:09Z",
      "ApplicationType": "Unspecified",
      "Architecture": "x86_64",
      "Url": "https://github.com/bottlerocket-os/bottlerocket",
      "Summary": "Settings plugin for the aws-k8s variants"
    },
    {
      "Name": "shibaken",
      "Publisher": "bottlerocket-core-kit",
      "Version": "0.0",
      "Release": "1.1731694217.dc9aff2d.br1",
      "Epoch": "1",
      "InstalledTime": "2024-11-22T18:29:09Z",
      "ApplicationType": "Unspecified",
      "Architecture": "x86_64",
      "Url": "https://github.com/bottlerocket-os/bottlerocket",
      "Summary": "IMDS client and settings generator"
    },
...
```
</details>

What users will see in SSM inventory collection:
Screenshot 2024-11-25 at 5 08 57 PM

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by yeazelm — APPROVED** [review]

File: twoliter/embedded/rpm2img [hunk 1]
```diff
@@ -186,61 +186,66 @@ INVENTORY_DATA="$(jq --raw-output . <<<"${installed_rpms[@]}")"
...
-  # Convert the bash array of core kit packages to a JSON array.
-  CORE_KIT_LIST="$(\
+mapfile -t kits < <(jq -r ".kit[].name" "${EXTERNAL_KIT_METADATA_PATH}")
+for kit in "${kits[@]}"; do
+  KIT_VERSION=$(jq --arg kit "${kit}" --raw-output '.kit[]|select(.name == $kit)|.version' "${EXTERNAL_KIT_METADATA_PATH}")
+  KIT_VENDOR=$(jq --arg kit "${kit}" --raw-output '.kit[]|select(.name == $kit)|.vendor' "${EXTERNAL_KIT_METADATA_PATH}")
+  # Set the path inside the build container to the kit RPMs and repo.
+  KIT_PATH="${EXTERNAL_KITS_PATH}/${KIT_VENDOR}/${kit}/${ARCH}"
+
+  if [[ -n "${KIT_VERSION}" && -n "${KIT_VENDOR}" ]]; then
```
Comment (cbgbt) [inline twoliter/embedded/rpm2img]: I think we should intentionally fail the build if this check fails.

Comment (ginglis13) [reply twoliter/embedded/rpm2img]: Good point, I added two distinct checks w/ error messages if these values are unset.

**Review by yeazelm — APPROVED** [review]

File: twoliter/embedded/rpm2img [hunk 2]
```diff
@@ -186,61 +186,73 @@ INVENTORY_DATA="$(jq --raw-output . <<<"${installed_rpms[@]}")"
...
+    > kit-replacements.json
 
-  # For any packages in the installed app inventory that exist in the core kit, replace
-  # the version with the core kit's version, and replace the name with the unprefixed name.
+  # For any packages in the installed app inventory that came from the kit, replace
+  # the Publisher with the kit's vendor and replace the name with the unprefixed name.
   INVENTORY_DATA="$(jq \
-    --argfile replace core-kit-replacements.json \
-     '.Content[].Name |= (if $replace[.] then $replace[.] else . end)' \
+    --argfile replace kit-replacements.json \
```
Comment (bcressey) [inline twoliter/embedded/rpm2img]: While you're here, can you get rid of this `--argfile` usage? It's apparently been deprecated for a long time and was removed in jq 1.7.

File: twoliter/embedded/rpm2img [hunk 3]
```diff
@@ -186,61 +186,73 @@ INVENTORY_DATA="$(jq --raw-output . <<<"${installed_rpms[@]}")"
...
+      echo "Failed to extract vendor of kit '${kit}'" >&2
+      exit 1
+  fi
+
+  # Set the path inside the build container to the kit RPMs and repo.
+  KIT_PATH="${EXTERNAL_KITS_PATH}/${KIT_VENDOR}/${kit}/${ARCH}"
+
+  # Query the kit repo for a single package's Release in order
+  # to extract the commit that built the kit.
+  KIT_GIT_SHA="$(dnf --repofrompath \
```
Comment (yeazelm) [inline twoliter/embedded/rpm2img]: Offline discussion helped us realize this isn't used anymore.

Comment (ginglis13) [reply twoliter/embedded/rpm2img]: https://github.com/bottlerocket-os/twoliter/commit/44a2623a6419dc7b07e346175a425bcad2f88aa6 switched to pulling package info directly from the RPM itself, so we can remove this entirely :D

**Comment by ginglis13** [comment]
^ force push replaces deprecated `--argfile` jq flag with `--slurpfile` and its slight syntax difference; also, remove `KIT_GIT_SHA` unused/dead code.

File: twoliter/embedded/rpm2img [hunk 4]
```diff
@@ -186,61 +186,55 @@ INVENTORY_DATA="$(jq --raw-output . <<<"${installed_rpms[@]}")"
...
-    --queryformat '%{Buildtime} %{Release}' 2>/dev/null | \
-    sort -k 1,2 | \
-    awk -F '.' 'END {print $--NF}')"
-  if [[ -z "${CORE_KIT_GIT_SHA}" ]]; then
-    echo "Could not find Git sha for bottlerocket-core-kit" >&2
-    exit 1
+mapfile -t kits < <(jq -r ".kit[].name" "${EXTERNAL_KIT_METADATA_PATH}")
+for kit in "${kits[@]}"; do
+  KIT_VERSION=$(jq --arg kit "${kit}" --raw-output '.kit[]|select(.name == $kit)|.version' "${EXTERNAL_KIT_METADATA_PATH}")
+  KIT_VENDOR=$(jq --arg kit "${kit}" --raw-output '.kit[]|select(.name == $kit)|.vendor' "${EXTERNAL_KIT_METADATA_PATH}")
```
Comment (bcressey) [inline twoliter/embedded/rpm2img]: style nit: for variables declared and used within a loop, I like to use lowercase variables as a gesture at proper scoping.

File: twoliter/embedded/rpm2img [hunk 5]
```diff
@@ -186,61 +186,55 @@ INVENTORY_DATA="$(jq --raw-output . <<<"${installed_rpms[@]}")"
...
-  # the version with the core kit's version, and replace the name with the unprefixed name.
+  # For any packages in the installed app inventory that came from the kit, replace
+  # the Publisher with the kit's vendor and replace the name with the unprefixed name.
   INVENTORY_DATA="$(jq \
-    --argfile replace core-kit-replacements.json \
-     '.Content[].Name |= (if $replace[.] then $replace[.] else . end)' \
+    --slurpfile replace kit-replacements.json \
+    --arg KIT "${kit}" \
+    '(.Content[] | select(.Name | $replace[][.] != null) | .Publisher) = $KIT |
+    .Content[].Name |= (if $replace[][.] then $replace[][.] else . end)' \
```
Comment (bcressey) [inline twoliter/embedded/rpm2img]: nit: indent this line a bit more to reflect that it's part of a continuing pipeline

**Review by bcressey — APPROVED** [review]

**Comment by ginglis13** [comment]
^ force push addresses some style feedback

**Review by cbgbt — APPROVED** [review]

**Pull request merged**